File: DSA/MySorting.py
from typing import List


class MySorting:
    def __init__(self) -> None:
        pass

    def selectionSort(self, nums):
        n = len(nums)
        for i in range(0, n):
            min_idx = i
            for j in range(i + 1, n):
                if nums[j] < nums[min_idx]:
                    min_idx = j
            if min_idx != i:
                nums[i], nums[min_idx] = nums[min_idx], nums[i]

    def bubbleSort(self, nums):
        n = len(nums)
        for i in range(0, n - 1):
            for j in range(0, n - i - 1):
                if nums[j] > nums[j + 1]:
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]

    def optimizedBubbleSort(self, nums):
        n = len(nums)
        for i in range(0, n - 1):
            flag = True
            for j in range(0, n - i - 1):
                if nums[j] > nums[j + 1]:
                    flag = False
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]
            if flag:
                break

    def inserttionSort(self, nums):
        n = len(nums)
        for i in range(0, n):
            j = i
            while j > 0 and nums[j - 1] > nums[j]:
                nums[j - 1], nums[j] = nums[j], nums[j - 1]
                j -= 1

    def __merge(self, nums: List[int], left: int, mid: int, right: int):
        n = right - left + 1
        mergedArr = [0] * n
        pt1, pt2, j = left, mid + 1, 0
        while pt1 <= mid and pt2 <= right and j < n:
            if nums[pt1] <= nums[pt2]:
                mergedArr[j] = nums[pt1]
                j += 1
                pt1 += 1
            else:
                mergedArr[j] = nums[pt2]
                j += 1
                pt2 += 1
        while pt1 <= mid and j < n:
            mergedArr[j] = nums[pt1]
            j += 1
            pt1 += 1
        while pt2 <= right and j < n:
            mergedArr[j] = nums[pt2]
            j += 1
            pt2 += 1
        nums[left : right + 1] = mergedArr

    def mergeSort(self, nums: List[int], left: int, right: int):
        if left >= right:
            return
        mid = (right + left) // 2
        self.mergeSort(nums, left, mid)
        self.mergeSort(nums, mid + 1, right)
        self.__merge(nums, left, mid, right)

    def __partition(self, nums: List[int], low: int, high: int) -> int:
        pivot = nums[low]
        start = low
        end = high
        while start < end:
            while nums[start] <= pivot and start <= high - 1:
                start += 1
            while nums[end] > pivot and end >= low + 1:
                end -= 1
            if start < end:
                nums[start], nums[end] = nums[end], nums[start]
        nums[end], nums[low] = nums[low], nums[end]
        return end

    def quickSort(self, nums: List[int], low: int, high: int):
        if low >= high:
            return
        pivot = self.__partition(nums, low, high)
        self.quickSort(nums, low, pivot - 1)
        self.quickSort(nums, pivot + 1, high)


x = MySorting()

# nums = [5, 4, 3, 2, 1]
# x.selectionSort(nums)
# print(nums)

# nums = [5, 4, 3, 2, 1]
# x.bubbleSort(nums)
# print(nums)

# nums = [5, 4, 3, 2, 1]
# x.optimizedBubbleSort(nums)
# print(nums)

# nums = [5, 4, 3, 2, 1]
# x.inserttionSort(nums)
# print(nums)

# nums = [5, 4, 3, 2, 1]
# x.mergeSort(nums, 0, len(nums) - 1)
# print(nums)

nums = [5, 4, 3, 2, 1]
x.quickSort(nums, 0, len(nums) - 1)
print(nums)

File: DSA/MySorting.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

class MySorting {
   private:
    void merge(vector<int> &arr, int low, int mid, int high) {
        const int n = high - low + 1;
        int *temp = new int[n];
        int pt1 = low, pt2 = mid + 1, j = 0;
        while (pt1 <= mid && pt2 <= high && j < n) {
            if (arr[pt1] < arr[pt2]) {
                temp[j++] = arr[pt1++];
            } else {
                temp[j++] = arr[pt2++];
            }
        }
        while (pt1 <= mid && j < n) {
            temp[j++] = arr[pt1++];
        }
        while (pt2 <= high && j < n) {
            temp[j++] = arr[pt2++];
        }
        j = 0;
        for (int i = low; i <= high; i++) {
            arr[i] = temp[j++];
        }
    }

    int partition(vector<int> &arr, int low, int high) {
        int pivotEle = arr[low];
        int start = low, end = high;
        while (start < end) {
            while (start <= high - 1 && arr[start] <= pivotEle) {
                start++;
            }
            while (end >= low + 1 && arr[end] > pivotEle) {
                end--;
            }
            if (start < end) {
                swap(arr[start], arr[end]);
            }
        }
        swap(arr[low], arr[end]);
        return end;
    }

   public:
    void selectionSort(vector<int> &arr) {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            int min_idx = i;
            for (int j = i + 1; j < n; j++) {
                if (arr[j] < arr[min_idx]) {
                    min_idx = j;
                }
            }
            if (min_idx != i) {
                swap(arr[i], arr[min_idx]);
            }
        }
    }

    void bubbleSort(vector<int> &arr) {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    swap(arr[j], arr[j + 1]);
                }
            }
        }
    }

    void optimizedBubbleSort(vector<int> &arr) {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            bool check = true;
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    check = false;
                    swap(arr[j], arr[j + 1]);
                }
            }
            if (check) {
                break;
            }
        }
    }

    void insertionSort(vector<int> &arr) {
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            int j = i;
            while (j > 0 && arr[j] < arr[j - 1]) {
                swap(arr[j], arr[j - 1]);
                j--;
            }
        }
    }

    void mergeSort(vector<int> &arr, int low, int high) {
        if (low >= high) {
            return;
        }
        int mid = (low + high) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, mid, high);
    }

    void quickSort(vector<int> &nums, int low, int high) {
        if (low >= high) {
            return;
        }
        int pivot = partition(nums, low, high);
        quickSort(nums, low, pivot - 1);
        quickSort(nums, pivot + 1, high);
    }

    void cyclicSort(vector<int> &nums) {
        int n = nums.size();
        int i = 0;
        while (i < n) {
            if (nums[i] - 1 == i) {
                i++;
            } else {
                swap(nums[i], nums[nums[i] - 1]);
            }
        }
    }

    // 268. Missing Number,
    // https://leetcode.com/problems/missing-number/submissions/1257064517/
    // 448. Find All Numbers Disappeared in an Array
    // https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/submissions/1257074682/
    // 287. Find the Duplicate Number
    // https://leetcode.com/problems/find-the-duplicate-number/submissions/1257230928/
    // 442. Find All Duplicates in an Array
    // https://leetcode.com/problems/find-all-duplicates-in-an-array/submissions/1257234180/
    // 645. Set Mismatch
    // 41. First Missing Positive
    // https://leetcode.com/problems/first-missing-positive/submissions/1257257377/
};

void printArr(vector<int> &nums) {
    int n = nums.size();
    for (int i = 0; i < n; i++) {
        cout << nums[i] << " - ";
    }
}

int main() {
    MySorting sort;

    vector<int> nums = {5, 4, 3, 2, 1};
    // sort.selectionSort(nums);

    // sort.bubbleSort(nums);

    // sort.optimizedBubbleSort(nums);

    // sort.insertionSort(nums);

    // sort.mergeSort(nums, 0, nums.size() - 1);

    // sort.quickSort(nums, 0, nums.size() - 1);

    sort.cyclicSort(nums);

    printArr(nums);
    return 0;
}
